src/main/java/uk/gov/ch/developer/docs/controller/developer/UserCallbackController.java

```diff
@@ -31,36 +26,21 @@
     private IOauth oauth;
 
     @GetMapping
-    public String getCallback(@RequestParam("state") String state,
-            @RequestParam("code") String code, final HttpServletRequest httpServletRequest) {
+    public void getCallback(@RequestParam("state") String state, @RequestParam("code") String code,
+            final HttpServletResponse httpServletResponse) {
         LOGGER.trace("Code:" + code);
         LOGGER.trace("State:" + state);
-        final Session chSession = (Session) httpServletRequest
-                .getAttribute(SessionHandler.CHS_SESSION_REQUEST_ATT_KEY);
-
-        final String returnedNonce = getNonceFromState(state);
-        if (!oauth.oauth2VerifyNonce(returnedNonce)) {
-            LOGGER.error("Invalid nonce value in state during oauth2 callback");
-            // return "redirect:/"; TODO redirect will not work, needs to be addressed for unmatched
-            // Nonce values
-        }
-
-        LOGGER.debug("Getting User Profile");
-
-        UserProfileResponse userProfileResponse = oauth.getUserProfile(code, chSession);
-
-        if (userProfileResponse == null) {
-            // TODO raise error
+        try {
+            final boolean valid = oauth.isValid(state, code);
+            if (valid) {
+                httpServletResponse.sendRedirect(identityProvider.getRedirectUriPage());
+            } else {
+                httpServletResponse.sendError(HttpServletResponse.SC_FORBIDDEN);
+            }
+        } catch (final Exception e) {
+            LOGGER.error(e);
+            httpServletResponse.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
```

Comment:
Resolved based upon `sendError` which is a better way of dealing with the 403 exceptions, however for a 500 we must remember to deal with this in the eventual global exception handler.

Reply:
Set status is used because send error can throw IOExceptions which we have to add to the signature.
If we do want a global controller advice to handle this we should replace this with a throw of a custom exception.